### PR TITLE
Cloudbot user management

### PR DESCRIPTION
**Pull request by AnkhHeart**
- Added User Management / Regulars
- Added custom Permission control so users can pick a combination of roles
- Added alert on duplicate regular adds / edits
- Updated beaker to 0.3.56
- Fixed a bug in regards to commands not showing up in the latest preview
- Wrong video command made available for Media Share

### CODE REVIEW

File: app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+      return;
+    }
+
+    this.previousCheckBoxes = cloneDeep(newValue);
+
+    value += newValue.subscriber.checked ? 2 : 0;
+    value += newValue.regular.checked ? 4 : 0;
+    value += newValue.moderator.checked ? 32 : 0;
+    value += newValue.streamer.checked ? 64 : 0;
```
Comment (andycreeth) [inline app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts]: There's an enum defined later in this PR.  Why isn't it being used?

File: app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        return type.checked;
+      });
+
+      this.text = map(checked, type => type.string).join(', ');
+    }
+
+    this.$emit('input', value);
+  }
+
+  updateChecks() {
```
Comment (andycreeth) [inline app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts]: Same comment for this function.

File: app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  toggle(key: string, value: boolean) {
+    this.checkboxes[key].checked = !value;
+  }
+
+  documentClick(e: MouseEvent) {
+    const el = this.$refs.paneMenu;
+    const target = e.target;
+
+    // @ts-ignore
+    if (el !== target && !el.contains(target)) {
```
Comment (andycreeth) [inline app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts]: You shouldn't need to disable TS if you declare the ref properly on your component.  Something like:
```TypeScript
$refs: {
  paneMenu: HTMLDivElement;
};
```

File: app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    window.removeEventListener('resize', this.onResize);
+  }
+
+  toggleDropDown() {
+    this.paneMenuOpen = !this.paneMenuOpen;
+    this.onResize();
+  }
+
+  onResize(event?: UIEvent) {
+    // @ts-ignore
```
Comment (andycreeth) [inline app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts]: Same comment as above.

File: app/services/chatbot/chatbot-interfaces.ts [hunk 5]
```diff
@@ -819,16 +838,28 @@ export interface IBettingCommand {
 // dictionaries
 export enum ChatbotAutopermitEnums {
   'None' = 0,
-  'Subscriber Only' = 1 << 1,
+  'Subscriber Only' = 2,
+  'Regulars Only' = 4,
+  'Regulars & Subscribers' = 6,
```
Comment (andycreeth) [inline app/services/chatbot/chatbot-interfaces.ts]: I think it's a little more clear if you keep the bit shifting here:
```TypeScript
export enum ChatbotAutopermitEnums {
  None = 0,
  SubscriberOnly = 1 << 1,
  RegularsOnly = 1 << 2,
  RegularsAndSubscribers = SubscriberOnly | RegularsOnly,
}
```

File: app/services/chatbot/chatbot-interfaces.ts [hunk 6]
```diff
@@ -819,16 +838,28 @@ export interface IBettingCommand {
...
+  'Subscribers, Regulars' = 6,
+  'Subscribers, Moderators' = 34,
+  'Subscribers, Streamer' = 66,
+  'Regulars, Moderators' = 36,
+  'Regulars, Streamer' = 68,
+  'Moderators, Streamer' = 96,
+  'Regulars, Subscribers, Moderators' = 38,
+  'Subscribers, Regulars, Streamer' = 70,
+  'Subscribers, Moderators, Streamer' = 98,
+  'Regulars, Moderators, Streamer' = 100,
```
Comment (andycreeth) [inline app/services/chatbot/chatbot-interfaces.ts]: Same comment as above.  This is not clear.  Also I don't think we should be using string names with unusual characters for these enum names, as it makes them more difficult to use.

Comment (AnkhHeart) [reply app/services/chatbot/chatbot-interfaces.ts]: Here it is possible unlike the ChatbotAutoPermitEnums, so done.

File: app/services/chatbot/chatbot-interfaces.ts (hunk 5 above)
Comment (AnkhHeart) [reply app/services/chatbot/chatbot-interfaces.ts]: Wouldn't work too well in this case. ChatbotAutoPermitEnums is used in the Auto permit dropdowns in the Mod Tools section.

File: app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts (hunk 4 above)
Comment (AnkhHeart) [reply app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts]: Done

File: app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts (hunk 3 above)
Comment (AnkhHeart) [reply app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts]: Done

File: app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts (hunk 2 above)
Comment (AnkhHeart) [reply app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts]: Done

File: app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts (hunk 1 above)
Comment (AnkhHeart) [reply app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts]: Done

File: app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts [hunk 7]
```diff
@@ -76,10 +89,14 @@ export default class ChatbotAliases extends ChatbotBase {
       this.checkboxes.streamer.checked = true;
       this.text = 'Everyone';
     } else {
-      this.checkboxes.subscriber.checked = (this.value & 2) === 2;
-      this.checkboxes.regular.checked = (this.value & 4) === 4;
-      this.checkboxes.moderator.checked = (this.value & 32) === 32;
-      this.checkboxes.streamer.checked = (this.value & 64) === 64;
+      this.checkboxes.subscriber.checked =
+        (this.value & ChatbotPermissionsEnums.Subscribers) === ChatbotPermissionsEnums.Subscribers;
```
[the remark this answers is not in the document]
Comment (AnkhHeart) [reply app/components/page-components/Chatbot/shared/ChatbotPermissions.vue.ts]: Done

**Pull request merged**